text/0000-js-compatible-resolution.md

```diff
@@ -0,0 +1,344 @@
+
+We can provide a shim utility to addon authors that allows them to reexport their MU components & helpers as a classic app tree, when the consuming app is not MU-aware. This lets addons upgrade unilterally, while clearly segregating the compatibility code within the shim utility, where it can be easily turned off in apps that don't need it.
+
+# How We Teach This
+
+This is intended to simplify the teaching story because
+
+ - users already need to learn how `import` works
+ - we are being careful to make sure it means exactly the same thing in templates
+
+
+# Drawbacks
+
+Explicit imports require more typing. A large template that uses a large number of components and helpers could end up with a big stack of imports. This is not a unique problem to our system, however, and it's already something developers need to learn to manage wisely in Javascript.
+
+This proposal does not require any breaking changes to Handlebars syntax (a Handlebars parser that doesn't know anything about it will treat the front matter as "just content", which is harmless). But if we want nice syntax highlighting and completion inside the frontmatter, we may need to update tools to be aware of it. On the plus side, once this work is done it leaves us in excellent position to expand the frontmatter to full single-file components.
+
+# Alternatives
+
+## Why "foo/index.js" over "foo/component.js"?
+
+Because `index.js` is well-understood by Javacript ecosystem tooling to be a file that will be found when you try to resolve `./foo`, whereas `foo/component.js` is not.
+
+We get less to customize and less to teach if we stick with the better supported word. And we loose very little by picking a different constant word.
+
+## Why "foo/index.hbs" over "foo/component.hbs"?
+
+This is _almost_ the same argument as the previous point, but with one caveat.
+
+It's true that when a Javascript file for the component exists, we are free to choose our own convention for which template to associate. So we _could_ make this work:
+
+ - foo/index.js
+ - foo/template.hbs
+
+However, this creates a refactoring hazard if the developer later deletes `foo/index.js`. This is _not_ resolvable by itself using standard node_modules rules:
+
+ - foo/template.hbs
+
+Whereas `foo/index.hbs` is resolvable, as long as we've registered a handler for `.hbs` files.
+
+# Unresolved questions
+
+It would probably be good to synthesize all the remaining-valid parts of the two earlier RFCs into this document, so there is one authoritative resource.
+
+This design leaves routes and their templates unchanged from the prior MU RFC. That seems to be an uncanny gap. They're still named `template.hbs`, which is maybe OK, but then if you want to nest more components under them you still need to reserve a place under `-components` to avoid ambiguity with child routes, and that means your imports in the route template need to say `import thing from "./-components/thing"` which seems bad.
```

Comment:
IMHO, we cannot get away without changing routes in MU design. Being forced to teach two very different systems is just too much. IMHO there is no real reason a route can't use the same unobtrusive association setup that was described above for both its controller and template.

Either way, doubling the teaching angle (how do I name xyz template) seems like a show stopper here to me...